### DISCUSSION TITLE
Handling of ccmp

### DISCUSSION THREAD

**Discussion by madig** [harfbuzz/harfbuzz#2856] [Q&A] [answered]
Hi!
I'm trying to test the `ccmp` feature of a font by writing text strings and comparing HB's output to what I believe should happen. I noticed that HB seems to prefer precomposed glyphs for sequences like `i + acutecomb` = `iacute`, so my tests have to use something not precomposed instead if I want to test `idotless` compositions. I'm trying to understand the process here and found https://github.com/harfbuzz/harfbuzz/blob/fd8f4ff/src/hb-ot-shape-normalize.cc#L52-L70 -- what does it mean for `ccmp` to be adequate? Is there more to it than what I highlighted? Is the behavior of other shaping engines known?

**Comment by behdad** [comment, marked as answer]
> what does it mean for `ccmp` to be adequate?

In that sentence it just meant if `ccmp` doesn't compose the decomposed form into the composed form. You can ignore what we are doing and as long as your font renders the various canonically-equivalent sequences the same, you should not be surprised by what HB does.

> Is there more to it than what I highlighted?

Not really. There have been fonts in the wild where for example the pre-composed forms do not have proper anchors for further mark attachments. Or that anchor placement produces different results for canonically-equivalent sequences. If your font does that, then it's not Unicode-compliant; HarfBuzz still tries hard to make it Unicode-compliant by picking one of those outputs and trying to always produce that for all canonically-equivalent sequences. This seems to surprise some people (https://github.com/harfbuzz/harfbuzz/issues/2832). But inherently is a font bug.

> Is the behavior of other shaping engines known?

As far as I know, no other shaper does things like this.

**Reply by madig** [reply]
Thanks for the details! Never heard of canonically-equivalent before, more stuff to look up.

> As far as I know, no other shaper does things like this.

As in, for other shaping engines, I really need an "adequate" `ccmp`?

**Reply by behdad** [reply]
> Thanks for the details! Never heard of canonically-equivalent before, more stuff to look up.

Is really simple. Read the Unicode Normalization Algorithm and understand it. Then check what HarfBuzz is doing and see how it makes sense.

> As in, for other shaping engines, I really need an "adequate" `ccmp`?

Is up to your font to implement canonical-equivalence. That means, you (font developer) have to make sure the precomposed, decomposed, and for sequences with more than one mark, any ordering of those marks, or even a precomposed followed by a mark, etc, all work correctly. One way to do that is to always decompose, or always compose, in the `ccmp` feature. But you can definitely get away without that as well, by just having correct anchors and correct mark skipping logic in your `mark` and `mkmk` features.

**Reply by madig** [reply]
One day I will read it. Random thought: make a font fuzzer that tests canonical equivalence by blasting it with Unicode sequences and testing that HarfBuzz spits out the same result with and without the automatism. If there's one thing that I learned over the years, it's that you can't trust fonts to do the right thing.
